tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/AWS/Shared/SpanPointers.cs

```diff
@@ -0,0 +1,95 @@
+// <copyright file="SpanPointers.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Security.Cryptography;
+using System.Text;
+
+namespace Datadog.Trace.ClrProfiler.AutoInstrumentation.AWS.Shared
+{
+    /// <summary>
+    ///  SpanPointer helper methods
+    /// </summary>
+    internal static class SpanPointers
+    {
+        // The pointer direction will always be down. The serverless agent handles cases where the
+        // direction is up.
+        private const string DownDirection = "d";
+        private const string LinkKind = "span-pointer";
+        private const int SpanPointerHashSizeBytes = 16;
+        private const string S3PtrKind = "aws.s3.object";
+
+        // S3 hashing rules: https://github.com/DataDog/dd-span-pointer-rules/blob/main/AWS/S3/Object/README.md
+        public static void AddS3SpanPointer(Span span, string bucketName, string key, string eTag)
+        {
+            var components = new[] { bucketName, key, StripQuotes(eTag) };
+            var hash = GeneratePointerHash(components);
+            var spanLinkAttributes = new List<KeyValuePair<string, string>>
+            {
+                new("ptr.kind", S3PtrKind),
+                new("ptr.dir", DownDirection),
+                new("ptr.hash", hash),
+                new("link.kind", LinkKind),
+            };
+
+            var spanLink = new SpanLink(SpanContext.ZeroContext, spanLinkAttributes);
+            span.AddLink(spanLink);
+        }
+
+        // Hashing rules: https://github.com/DataDog/dd-span-pointer-rules/tree/main?tab=readme-ov-file#general-hashing-rules
+        private static string GeneratePointerHash(string[] components)
+        {
+            using var stream = new MemoryStream();
```

Comment:
[perf] Like the `List<>` capacity above, if you know the size you should provide it in the `MemoryStream` constructor. You can probably even use a `byte[]` directly instead of allocating a `MemoryStream`. You can go even further in .NET Code and allocate that `byte[]` on the stack if it's small enough.

It's harder to write the code, but we try to be a little extreme for code that's in the hot path. 😅

Reply:
I pushed commit to try to minimize allocations here. There's differences for the runtimes we target, but even for .NET Framework we can do better. Hope you don't mind!